### ISSUE TITLE
Investigate replacing compress/gzip with klauspost/compress

### ISSUE THREAD

**Issue by cmacknz** [elastic/beats#38599]
The APM Server recently observed a +10% more throughput with a -15% reduction in CPU usage after switching from `compress/gzip` to https://github.com/klauspost/compress for gzip compression when indexing to _bulk.

Make the same change in Beats an experimentally verify that we get the same results.

**Comment by khushijain21** [comment]
We indeed see a drop in cpu profiles with `klauspost/compress`. The cpu utilization of `bulkEncode` method decreased from 11% to 6%.

 But the gains are not reflected on the overall metrics. [here](https://ingest-metricstore.kb.us-central1.gcp.cloud.es.io:9243/app/dashboards#/view/ce699545-7f3f-42ce-b0f9-16c51e8b01ef?_g=(refreshInterval:(pause:!t,value:60000),time:(from:'2024-11-08T10:43:14.259Z',to:'2024-11-08T11:53:17.764Z')))

Overall CPU utilization only drops from 36% to 35%. 
And bytes written to output reduced from `639MB` to `592MB` - which improves the network traffic here.

**Comment by cmacknz** [comment]
Yes, it is used in https://github.com/elastic/go-docappender/blob/7a214a4fba74f4835375dbde2ce7499ee657d18a/bulk_indexer.go#L32 which is in turn used in the OTel collector's Elasticsearch exporter. Making this the same removes a possible source of performance difference.

The difference is smaller than I hoped here but you can get larger gains over time by stacking together small wins like this, so I think we should still move forward.

**Comment by khushijain21** [comment]
so on re-running the benchmark with varying compression levels we have the following overall metric

- We do not see an overall cpu gain. EPS and other metrics vary by 2%. 
- The bytes written to o/p vary as follows. We see an increase in traffic with higher compression levels and no overall cpu gain. 

| Test                | without_klauspost | with_klauspost | Network Traffic |
| ------------------- | ----------------- | -------------- | --------------- |
| Compression Level 1 | 745               | 693            | -7%             |
| Compression Level 4 | 565               | 658            | +16.46%         |
| Compression Level 8 | 453               | 467            | +3.09%          |

There is a tradeoff here between cpu utilization and network traffic as we move (if) towards `klauspost/compress` library.

**Comment by rdner** [comment]
@cmacknz It does not look we're gaining anything on the stack level. We're just moving CPU load to the network load. I have concerns whether it's the right call to switch to `klauspost`.

@khushijain21 is it possible to identify at which compression levels both libraries produce a payload about the same size with optimal CPU usage?

I think only comparing these 2 libraries at the same output size we can isolate the CPU gain, so we can compare the libraries fairly. Otherwise, it seems like we moving load elsewhere and by this achieving less load on the CPU.

**Comment by rdner** [comment]
I reviewed the benchmark results with @khushijain21 and it's clear that we get some gains from `klauspost`.

My concern is more about existing presets (do they need adjustment, if yes then how?) and perhaps impact on pricing when continue to use the same presets.

It can be beyond the scope of this issue though.
